### ISSUE TITLE
Bad digits with emoji fonts

### ISSUE THREAD

**Issue by mhosken** [Kozea/WeasyPrint#1354]
This sample:

```
<html><header/><body>
<p style="font-family: sans">U+0031, U+0E41</p>
<p style="font-family: sans, &quot;Noto Color Emoji&quot;">U+0031, U+0E41</p>
</body></html>
```
Renders the first line correctly, but messes up the second line badly, with evince complaining `some font thing failed`.

I suspect a problem with Pango, but I can't see where. Digits are keycap characters, but only keycap characters with zwj between or with a VS16 are supposed to be considered emoji characters.

**Comment by liZe** [comment]
> with evince complaining `some font thing failed`

Hello!

Do you use Cairo 1.17.4? If so, this bug is a duplicate of #1292 that you can fix by downgrading Cairo or using a patched version of Cairo (some Linux distributions already have the patch in their latest version).

**Comment by mhosken** [comment]
I don't think so. I am running libcairo 1.16.0 as found in Ubuntu 20.04. It also fails on latest MacOS with no complaints from its PDF viewer.

What's happening is that Pango is saying the digits in the text need to come from the Emoji font and are emoji and yet should be coming from the main sans font. As stated before, I can find nothing in the Pango source that would imply that this has to happen since the rules say you have to have zwj and VS16 in the text string. The resulting PDF has both the sans and the Emoji fonts in and switches between them in the page output, but the Emoji font has not digit characters (obviously) hence the problematic rendering.

**Comment by liZe** [comment]
I don’t have the error in Evince. Here is the PDF I get:

[emoji.pdf](https://github.com/Kozea/WeasyPrint/files/6546457/emoji.pdf)

Numbers of the second line are emojis from Noto. The Noto Emoji font has these numbers, that’s where the bitmap numbers come from.

> I suspect a problem with Pango, but I can't see where. Digits are keycap characters, but only keycap characters with zwj between or with a VS16 are supposed to be considered emoji characters.

Maybe we don’t use Pango as we should.

Even with the current master branch (where Cairo has been removed and colored emojis don’t work), we have emojis (actually empty spaces because it’s broken) instead of normal numbers. At least we know that it’s not a problem with Cairo.

**Comment by liZe** [comment]
Must be fixed in #1406.
